From 4e3ac24708f465235e04c7b377fb4d86dbd86347 Mon Sep 17 00:00:00 2001
From: Tiberiu Popescu <tiberiu1204@gmail.com>
Date: Wed, 15 Oct 2025 22:33:29 +0300
Subject: [PATCH 17/27] Added option to send vie tcp to mprotect hook

---
 security/mcheck/mcheck.c | 158 ++++++++++++++++++++-------------------
 1 file changed, 83 insertions(+), 75 deletions(-)

diff --git a/security/mcheck/mcheck.c b/security/mcheck/mcheck.c
index 7570f8346..0c743ac57 100644
--- a/security/mcheck/mcheck.c
+++ b/security/mcheck/mcheck.c
@@ -29,7 +29,7 @@
 
 // Module param to switch between netlink and tcp socket transport methods
 
-static bool use_tcp = false;
+static bool use_tcp = true;
 module_param(use_tcp, bool, 0644);
 MODULE_PARM_DESC(use_tcp, "Send messages to userspace via a TCP socket");
 
@@ -323,87 +323,92 @@ static int mmap_analyzer_thread(void *arg) {
     return 0;
 }
 
-static int mcheck_custom_mmap_hook(unsigned long addr, unsigned long len, unsigned long prot)
+static int send_pages_to_userspace_tcp(unsigned long addr, unsigned long len, unsigned long prot)
 {
-    // pr_info("mcheck_custom_mmap_hook: addr = %lx, len = %lx, prot = %lx\n", addr, len, prot);
-	
-	if (use_tcp) {
-		if (!sock) {
-			if (lsm_init_tcp_connection() != 0) {
-				// server cannot be reached, return 0
-				return 0;
-			}
+	if (!sock) {
+		if (lsm_init_tcp_connection() != 0) {
+			// server cannot be reached, return 0
+			return 0;
 		}
-        if (!addr || !len) {
-            pr_err("LSM: Invalid buffer or length for TCP send\n");
-            return -EINVAL;
-        }
-		// we are only interested in executable mappings
-		if (!(prot & PROT_EXEC)) {
-    	    return 0;
-    	}
-
-        unsigned long pages_no = (len + PAGE_SIZE - 1) >> PAGE_SHIFT;
-        struct page **pages = kmalloc(sizeof(struct page*) * pages_no, GFP_KERNEL);
-        if (!pages) {
-            pr_err("LSM: Failed to allocate page array\n");
-            return -ENOMEM;
-        }
+	}
+    if (!addr || !len) {
+        pr_err("LSM: Invalid buffer or length for TCP send\n");
+        return -EINVAL;
+    }
+	// we are only interested in executable mappings
+	if (!(prot & PROT_EXEC)) {
+	    return 0;
+	}
 
-        int pinned = get_user_pages(addr, pages_no, FOLL_FORCE | FOLL_WRITE, pages);
-        if (pinned <= 0) {
-            pr_err("LSM: Failed to pin pages\n");
-            kfree(pages);
-            return -EFAULT;
-        }
+    unsigned long pages_no = (len + PAGE_SIZE - 1) >> PAGE_SHIFT;
+    struct page **pages = kmalloc(sizeof(struct page*) * pages_no, GFP_KERNEL);
+    if (!pages) {
+        pr_err("LSM: Failed to allocate page array\n");
+        return -ENOMEM;
+    }
+
+    int pinned = get_user_pages(addr, pages_no, FOLL_FORCE | FOLL_WRITE, pages);
+    if (pinned <= 0) {
+        pr_err("LSM: Failed to pin pages\n");
+        kfree(pages);
+        return -EFAULT;
+    }
+
+	struct vm_area_struct *vma = find_vma(current->mm, addr);
+	if (!vma) {
+	    pr_err("mcheck_custom_mmap_hook: failed to find vma\n");
+	    return -EINVAL;
+	}
 
-		struct vm_area_struct *vma = find_vma(current->mm, addr);
-		if (!vma) {
-    	    pr_err("mcheck_custom_mmap_hook: failed to find vma\n");
-    	    return -EINVAL;
-    	}
-
-       for (int i = 0; i < pinned; i++) {
-			void *kaddr = kmap(pages[i]);
-    		size_t chunk_size = (i == pinned - 1) ? (len - i * PAGE_SIZE) : PAGE_SIZE;
-
-    		struct mapping_info info = {
-    		    .pid = current->pid,
-    		    .init_addr = addr + i * PAGE_SIZE,  
-    		    .mapped_addr = 0,                   
-    		    .len = chunk_size,                  
-    		    .prot = prot,                       
-    		    .is_file_backed = vma->vm_file != NULL,                
-    		};
-
-    		char *buf = kmalloc(sizeof(info) + PAGE_SIZE, GFP_KERNEL);
-			if (!buf) {
-				pr_err("mcheck_custom_mmap_hook: kmalloc failed\n");
-				return -ENOMEM;
-			}
-
-    		memcpy(buf, &info, sizeof(info));
-    		memcpy(buf + sizeof(info), kaddr, chunk_size);
-
-    		lsm_send_to_userspace_tcp(buf, sizeof(info) + chunk_size);
-
-    		kunmap(pages[i]);
-    		put_page(pages[i]);
-			kfree(buf);
+   for (int i = 0; i < pinned; i++) {
+		void *kaddr = kmap(pages[i]);
+		size_t chunk_size = (i == pinned - 1) ? (len - i * PAGE_SIZE) : PAGE_SIZE;
+
+		struct mapping_info info = {
+		    .pid = current->pid,
+		    .init_addr = addr + i * PAGE_SIZE,  
+		    .mapped_addr = 0,                   
+		    .len = chunk_size,                  
+		    .prot = prot,                       
+		    .is_file_backed = vma->vm_file != NULL,                
+		};
+
+		char *buf = kmalloc(sizeof(info) + PAGE_SIZE, GFP_KERNEL);
+		if (!buf) {
+			pr_err("mcheck_custom_mmap_hook: kmalloc failed\n");
+			return -ENOMEM;
 		}
 
-		pr_info("LSM: Sent %ld pages to userspace\n", pages_no);
+		memcpy(buf, &info, sizeof(info));
+		memcpy(buf + sizeof(info), kaddr, chunk_size);
 
-        kfree(pages);
+		lsm_send_to_userspace_tcp(buf, sizeof(info) + chunk_size);
+
+		kunmap(pages[i]);
+		put_page(pages[i]);
+		kfree(buf);
+	}
+
+	pr_info("LSM: Sent %ld pages to userspace\n", pages_no);
+
+    kfree(pages);
+
+	// char buf[16];
+	// ssize_t len = lsm_receive_reply_tcp(buf);
+	// if (len < 0) {
+	// 	pr_err("LSM: Failed to receive analyzer reply");
+	// 	return -EINVAL;
+	// }
+	
+    return -analyzer_response;  // Early exit, skip analyzer mapping
+}
+
+static int mcheck_custom_mmap_hook(unsigned long addr, unsigned long len, unsigned long prot)
+{
+    // pr_info("mcheck_custom_mmap_hook: addr = %lx, len = %lx, prot = %lx\n", addr, len, prot);
 
-		// char buf[16];
-		// ssize_t len = lsm_receive_reply_tcp(buf);
-		// if (len < 0) {
-		// 	pr_err("LSM: Failed to receive analyzer reply");
-		// 	return -EINVAL;
-		// }
-		
-        return -analyzer_response;  // Early exit, skip analyzer mapping
+	if (use_tcp) {
+		return send_pages_to_userspace_tcp(addr, len, prot);
     }
 
     // analyzer task is not set yet
@@ -497,7 +502,10 @@ static int mcheck_custom_mmap_hook(unsigned long addr, unsigned long len, unsign
 static int mcheck_custom_mprotect_hook(unsigned long addr, unsigned long len, unsigned long prot)
 {
     //pr_info("mcheck_custom_mprotect_hook: addr = %lx, len = %lx, prot = %lx\n", addr, len, prot);
-
+	if (use_tcp) {
+		return send_pages_to_userspace_tcp(addr, len, prot);
+	}
+	
     // analyzer task is not set yet
     if (!analyzer_task) {
         return 0;
-- 
2.54.0

